### PR TITLE
LightningCLI

### PR DESCRIPTION
**Pull request by ordabayevy**
- Update examples to use `LightningCLI`
- Update `ModuleCheckpoint` to save `state_dict`. (Saving modules directly causes pickling issues)
- `ModuleCheckpoint` saves on `rank_zero_only` and every `log_every_n_steps`
- Attempt to read `total_variance` from the checkpoint in `VarianceMonitor`
- Add `DistributedAnnDataCollectionDataModule`
- Add `OnePassMeanVarStdFromCli` and `ProbabilisticPCAFromCli`
- Add `g_genes` argument to `OnePassMeanVarStd`

### CODE REVIEW

File: examples/onepass_mean_var_std.py [hunk 1]
```diff
@@ -9,76 +9,42 @@
...
-    )
-    data_loader = torch.utils.data.DataLoader(
-        dataset,
-        num_workers=args.num_workers,
-        collate_fn=collate_fn,
-    )
+class OnePassMeanVarStdCLI(LightningCLI):
+    """LightningCLI with custom argument linking."""
+
+    def add_arguments_to_parser(self, parser):
```
Comment (mbabadi) [inline examples/onepass_mean_var_std.py:35]: What's the type of `parser`?

File: examples/probabilistic_pca.py [hunk 2]
```diff
@@ -8,112 +8,59 @@
 This example shows how to fit feature count data to probabilistic PCA
 model [1].
 
-There are three flavors of probabilistic PCA model that are available:
+There are two flavors of probabilistic PCA model that are available:
```
Comment (mbabadi) [inline examples/probabilistic_pca.py:11]: Maybe write a word in the docstring about the pros and cons of the two approaches.

File: scvid/callbacks/module_checkpoint.py [hunk 3]
```diff
@@ -46,18 +47,27 @@ def setup(
...
-
+        if (
+            self.save_on_train_batch_end
+            and trainer.log_every_n_steps != 0  # type: ignore[attr-defined]
+            and trainer.global_step % trainer.log_every_n_steps == 0  # type: ignore[attr-defined]
+        ):
+            assert isinstance(pl_module.module, torch.nn.Module)
+            torch.save(pl_module.module.state_dict(), self.filepath)
+
+    @rank_zero_only
```
Comment (mbabadi) [inline scvid/callbacks/module_checkpoint.py:70]: Huh!  That's a nice decorator! :)

File: scvid/module/from_cli.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+import torch
+
+from scvid.transforms import ZScoreLog1pNormalize
+
+from ._probabilisticpca import ProbabilisticPCA
+from .onepass_mean_var_std import OnePassMeanVarStd
+
+
+class OnePassMeanVarStdFromCli(OnePassMeanVarStd):
```
Comment (mbabadi) [inline scvid/module/from_cli.py:14]: `OnePassMeanVarStdFromCli` -> `OnePassMeanVarStdFromCLI`

File: scvid/module/from_cli.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    """
+
+    def __init__(self, g_genes, target_count: int = 10_000) -> None:
+        transform = ZScoreLog1pNormalize(
+            mean_g=0, std_g=None, perform_scaling=False, target_count=target_count
+        )
+        super().__init__(g_genes, transform=transform)
+
+
+class ProbabilisticPCAFromCli(ProbabilisticPCA):
```
Comment (mbabadi) [inline scvid/module/from_cli.py:30]: `ProbabilisticPCAFromCli` -> `ProbabilisticPCAFromCLI`

File: scvid/module/from_cli.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+from .onepass_mean_var_std import OnePassMeanVarStd
+
+
+class OnePassMeanVarStdFromCli(OnePassMeanVarStd):
+    """
+    Preset default values for the LightningCLI.
+
+    Args:
+        g_genes: Number of genes.
+        target_count: Target gene epxression count. Default: ``10_000``.
```
Comment (mbabadi) [inline scvid/module/from_cli.py:20]: epxression -> expression

File: scvid/module/from_cli.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    """
+
+    def __init__(self, g_genes, target_count: int = 10_000) -> None:
```
Comment (mbabadi) [inline scvid/module/from_cli.py:23]: We must think of a mechanism to synchronize the `target_count` used here and the target count used by PCA modules.  At the very least, there must be an assert statement to ensure that mean and std that are precomputed here could only be used by a PCA module that is instantiated with the same `target_count`.

File: scvid/train/training_plan.py [hunk 8]
```diff
@@ -84,14 +84,15 @@ def configure_optimizers(self) -> dict[str, Any]:
...
+    def on_train_epoch_start(self) -> None:
         """
         Calls the ``set_epoch`` method on the iterable dataset of the given dataloader.
 
         If the dataset is ``IterableDataset`` and has ``set_epoch`` method defined, then
         ``set_epoch`` must be called at the beginning of every epoch to ensure shuffling
         applies a new ordering. This has no effect if shuffling is off.
         """
+        assert self.trainer.fit_loop._combined_loader is not None
         dataloaders = self.trainer.fit_loop._combined_loader.flattened
```
Comment (mbabadi) [inline scvid/train/training_plan.py:96]: Could you add some docstring about the logic here? What is `trainer.fit_loop._combined_loader` for those of us who don't know much PL? :p

**Review by mbabadi — CHANGES_REQUESTED** [review]
Nicely done!  The code looks much cleaner using LightningCLI.  I only requested a few small changes + documentation.

File: examples/onepass_mean_var_std.py [hunk 9]
```diff
@@ -9,76 +9,42 @@
...
-    )
-    dataset = IterableDistributedAnnDataCollectionDataset(
-        dadc, batch_size=args.batch_size, shuffle=True
-    )
-    data_loader = torch.utils.data.DataLoader(
-        dataset,
-        num_workers=args.num_workers,
-        collate_fn=collate_fn,
-    )
+class _LightningCLIWithLinks(LightningCLI):
```
Comment (ordabayevy) [inline examples/onepass_mean_var_std.py:32]: Changed to `_LightningCLIWithLinks`

File: examples/probabilistic_pca.py [hunk 10]
```diff
@@ -8,112 +8,66 @@
...
-    )
-    dataset = IterableDistributedAnnDataCollectionDataset(
-        dadc, batch_size=args.batch_size, shuffle=True
-    )
-    data_loader = torch.utils.data.DataLoader(
-        dataset,
-        num_workers=args.num_workers,
-        collate_fn=collate_fn,
-    )
+class _LightningCLIWithLinks(LightningCLI):
```
Comment (ordabayevy) [inline examples/probabilistic_pca.py:53]: _LightningCLIWithLinks

File: scvid/module/from_cli.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            # compute W_init_scale and sigma_init_scale
+            W_init_scale = W_init_variance_ratio
+            sigma_init_scale = sigma_init_variance_ratio
+            mean_g = None
+        else:
+            # load OnePassMeanVarStdFromCLI from checkpoint
+            onepass: OnePassMeanVarStdFromCLI = torch.load(mean_var_std_ckpt_path)
+            assert isinstance(onepass.transform, ZScoreLog1pNormalize)
+            assert target_count == onepass.transform.target_count
+            assert g_genes == onepass.g_genes
```
Comment (ordabayevy) [inline scvid/module/from_cli.py:72]: Added assertion statements here for `target_count` and `g_genes`.

File: scvid/train/training_plan.py [hunk 12]
```diff
@@ -84,15 +84,20 @@ def configure_optimizers(self) -> dict[str, Any]:
...
         Calls the ``set_epoch`` method on the iterable dataset of the given dataloader.
 
         If the dataset is ``IterableDataset`` and has ``set_epoch`` method defined, then
         ``set_epoch`` must be called at the beginning of every epoch to ensure shuffling
         applies a new ordering. This has no effect if shuffling is off.
         """
-        dataloaders = self.trainer.fit_loop._combined_loader.flattened
+        # dataloader is wrapped in a combined loader and can be accessed via
+        # flattened property which returns a list of dataloaders
+        # https://lightning.ai/docs/pytorch/stable/api/lightning.pytorch.utilities.combined_loader.html
```
Comment (ordabayevy) [inline scvid/train/training_plan.py:97]: Added comments here.

File: test/test_onepass_mean_var_std.py [hunk 13]
```diff
@@ -178,23 +178,28 @@ def test_module_checkpoint(tmp_path: Path, checkpoint_kwargs: dict):
...
     # load model from checkpoint
     assert os.path.exists(os.path.join(tmp_path, "module_checkpoint.pt"))
-    loaded_model = torch.load(os.path.join(tmp_path, "module_checkpoint.pt"))
+    loaded_model: OnePassMeanVarStdFromCLI = torch.load(
+        os.path.join(tmp_path, "module_checkpoint.pt")
+    )
     # assert
+    assert isinstance(model.transform, ZScoreLog1pNormalize)
+    assert isinstance(loaded_model.transform, ZScoreLog1pNormalize)
+    assert model.transform.target_count == loaded_model.transform.target_count
```
Comment (ordabayevy) [inline test/test_onepass_mean_var_std.py:202]: Added assertion for `target_count` of the loaded model.

File: test/test_ppca.py [hunk 14]
```diff
@@ -123,3 +125,54 @@ def test_variance_monitor(x_ng: np.ndarray):
...
+            "save_on_train_batch_end": False,
+        },
+        {
+            "save_on_train_end": False,
+            "save_on_train_epoch_end": False,
+            "save_on_train_batch_end": True,
+        },
+    ],
+)
+def test_module_checkpoint(tmp_path: Path, checkpoint_kwargs: dict):
```
Comment (ordabayevy) [inline test/test_ppca.py:150]: Added `ModuleCheckpoint` test for `ProbabilisticPCAFromCLI`.

**Review by ordabayevy — COMMENTED** [review]
@mbabadi thanks for reviewing!
Addressed your comments. Can you have another look and approve it?

File: scvid/module/from_cli.py (hunk 7 above)
Comment (ordabayevy) [reply scvid/module/from_cli.py:23]: I reverted back `ModuleCheckpoint` to save the model itself and not the `state_dict`. This way it saves `target_count`. Then we can use an assert statement for `target_count` when `ProbabilisticPCAFromCLI` is instantiated.

Another consideration here is that when the model is deployed we need to be able to load it and perform embedding. After #62 is merged the model (`IncrementalPCA`) can be saved with `ModuleCheckpoint` and then @fedorgrab should be able to load it and perform embedding with it.

**Review by mbabadi — APPROVED** [review]

**Pull request merged**